tests/spicy/types/unit/synchronized-nested.spicy

```diff
@@ -0,0 +1,47 @@
+# @TEST-DOC: Exercises synchronization on hierarchies with nesting.
+
+# @TEST-EXEC: spicyc -j -d %INPUT -o test.hlto
+
+# @TEST-EXEC: echo '# Base cases without synchronization' >>output
+# @TEST-EXEC: ${SCRIPTS}/printf 'X1' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC: ${SCRIPTS}/printf 'X1' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC: ${SCRIPTS}/printf 'X12' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC: ${SCRIPTS}/printf 'X12\x03' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC: ${SCRIPTS}/printf 'X12\x034' | spicy-dump -d test.hlto >>output 2>&1
+
+# @TEST-EXEC: echo '# Synchronize on non-vector field' >>output
+# @TEST-EXEC: ${SCRIPTS}/printf 'x1' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC: ${SCRIPTS}/printf 'x2' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC: ${SCRIPTS}/printf 'x\x03' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC: ${SCRIPTS}/printf 'x4' | spicy-dump -d test.hlto >>output 2>&1
+# @TEST-EXEC-FAIL: ${SCRIPTS}/printf 'x5' | spicy-dump -d test.hlto >>output 2>&1
```

Comment:
What happens here is that when trying to parse `X::x` which is `/X/` there is no match for an input `x5`. The parser then enters sync mode and tries to match against all possible values in `A` (`b"1"`, `/2/`, `uint8(3)`, `/4/`). Neither `x` nor `5 match here so synchronization fails. The parser then rethrows the parse error _which caused it to enter sync mode_ which is this case we the one from the failed match of `x` against `X::x`.

AFAICT this works as intended.

Reply:
Yep, you're right there. As discussed, extending the error with context about the failed sync attempt would be helpful.

Reply:
In order to augment the error string we need a way to get it in HILTI code. We can get that by e.g., adding support for formatting to `Exception` so we can call `"more context: %s" % exception`. The downside of that is that our exception classes already include the exception kind in their error string, so that rethrowing a `ParseError` with a reformatted error string from the original error would get us something like this:

```
[error] terminating with uncaught exception of type spicy::rt::ParseError: parse error: failed to synchronize: parse error: no expected look-ahead token found
```

I'll go that route and will punt on the stringification cleanup across all exceptions for now.

Reply:
I added a patch dropping the `parse error:` prefix from the `ParseError` altogether.